model/post.go

```diff
@@ -390,3 +401,76 @@ func (o *Post) GenerateActionIds() {
 		}
 	}
 }
+
+var markdownDestinationEscaper = strings.NewReplacer(
+	`\`, `\\`,
+	`<`, `\<`,
+	`>`, `\>`,
+	`(`, `\(`,
+	`)`, `\)`,
+)
+
+func (o *Post) WithRewrittenImageURLs(f func(string) string) *Post {
+	copy := *o
+	copy.Message = RewriteImageURLs(o.Message, f)
+	return &copy
+}
+
+func RewriteImageURLs(message string, f func(string) string) string {
+	if !strings.Contains(message, "![") {
+		return message
+	}
+
+	var ranges []markdown.Range
+
+	markdown.Inspect(message, func(blockOrInline interface{}) bool {
+		switch v := blockOrInline.(type) {
+		case *markdown.ReferenceImage:
+			ranges = append(ranges, v.ReferenceDefinition.RawDestination)
+		case *markdown.InlineImage:
+			ranges = append(ranges, v.RawDestination)
```

Comment:
No other Markdown library tracks enough information to actually be able to modify the source without making unnecessary changes to whitespace or syntax. You can inspect the AST with them... but you can't reassemble the original source or find out what source line / column an image URL was defined in.

I really wanted the output of this function to look _exactly_ like the input in every way possible, so I went through great care to make sure you can map block metadata back to source code ranges. It's kinda like the Go parsing packages, except a lot more straight-forward: `.RawDestination` simply provides the range of the raw, unescaped destination string in the original markdown.